llvm-19.1.7 %oneapi@2025.0.4 build fails: undefined reference to `_intel_fast_memcpy'

### Steps to reproduce the issue

Build of `llvm@19.1.7 %oneapi@2025.0.4` is failing using:
* Ubuntu 22.04, AMD64
* OneAPI 2025.0.4
* `spack@develop` (abd34875701a00c92ff67bf9fa13cc85e40d775e from `Thu Mar 20 09:27:11 2025 -0700`)

Spack environment:  [spack.yaml.txt](https://github.com/user-attachments/files/19375348/spack.yaml.txt)

<details><summary>Concretization</summary>

<pre>
 -   llvm@19.1.7+clang~cuda~flang+gold~ipo+libomptarget~libomptarget_debug~link_llvm_dylib+lld+lldb+llvm_dylib+lua~mlir+offload+polly~python~split_dwarf~z3~zstd build_system=cmake build_type=Release compiler-rt=runtime generator=ninja libcxx=runtime libunwind=runtime openmp=runtime shlib_symbol_version=none targets=all version_suffix=none arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^binutils@2.43.1~debuginfod~gas+gold~gprofng+headers~interwork+ld+libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs=shared,static arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^diffutils@3.10 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^gmake@4.4.1~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^zstd@1.5.6+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^cmake@3.31.6~doc+ncurses+ownlibs~qtgui build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^curl@8.11.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^nghttp2@1.65.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^openssl@3.4.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]                  ^ca-certificates-mozilla@2025-02-25 build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[e]      ^glibc@2.35 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^hwloc@2.11.1~cairo~cuda~gl~level_zero~libudev+libxml2~nvml~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^libpciaccess@0.17 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^util-macros@1.20.1 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^intel-oneapi-runtime@2025.0.4 build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^gcc-runtime@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %gcc@11.4.0
[+]      ^libedit@3.1-20240808 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^libffi@3.4.6 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^libxml2@2.13.5~http+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^libiconv@1.17 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^lua@5.3.6+shared build_system=makefile fetcher=curl arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^readline@8.2 build_system=autotools patches=1ea4349,24f587b,3d9885e,5911a5b,622ba38,6c8adf8,758e2ec,79572ee,a177edc,bbf97f1,c7b45ff,e0013d9,e065038 arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^unzip@6.0 build_system=makefile patches=881d2ed,f6f6236 arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^ncurses@6.5~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^ninja@1.12.1+re2c build_system=generic patches=93f4bb3 arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^re2c@3.1 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^perl-data-dumper@2.173 build_system=perl arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^perl@5.40.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^berkeley-db@18.1.40+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^bzip2@1.0.8~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]              ^gdbm@1.23 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^pkgconf@2.3.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[e]      ^python@3.10.12+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,ebdca64,f2fd060 arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^swig@4.1.1 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]          ^pcre2@10.44~jit+multibyte+pic build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^xz@5.6.3+pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
[+]      ^zlib-ng@2.2.3+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3 %oneapi@2025.0.4
</pre>

</details>

Error reproduced below in Docker container image:
* `ecpe4s/ubuntu22.04-runner-amd64-oneapi-2025.0:2025.03.01`

```
$> spack -e . install
...
==> Installing llvm-19.1.7-ggylpyfhgdtne32b4cqxsqpmmdyngqji [47/48]
==> No binary for llvm-19.1.7-ggylpyfhgdtne32b4cqxsqpmmdyngqji found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/59/59abea1c22e64933fad4de1671a61cdb934098793c7a31b333ff58dc41bff36c.tar.gz
==> No patches needed for llvm
==> llvm: Executing phase: 'cmake'
==> llvm: Executing phase: 'build'
...
     12028    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/oneapi-2025.0.4/binutils-2.43.1-6o5ttv7sqyrmss7z7yhjxw22pubvmuu7/bin/ld: /tmp/root/spack-stage/spack-stage-llvm-19.1.7
              -ggylpyfhgdtne32b4cqxsqpmmdyngqji/spack-build-ggylpyf/lib/libLLVMSupport.a(WithColor.cpp.o): in function `llvm::WithColor::error(llvm::raw_ostream&, llvm::StringRe
              f, bool)':
  >> 12029    WithColor.cpp:(.text._ZN4llvm9WithColor5errorERNS_11raw_ostreamENS_9StringRefEb+0x4c): undefined reference to `_intel_fast_memcpy'
     12030    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/oneapi-2025.0.4/binutils-2.43.1-6o5ttv7sqyrmss7z7yhjxw22pubvmuu7/bin/ld: /tmp/root/spack-stage/spack-stage-llvm-19.1.7
              -ggylpyfhgdtne32b4cqxsqpmmdyngqji/spack-build-ggylpyf/lib/libLLVMSupport.a(WithColor.cpp.o): in function `llvm::WithColor::warning(llvm::raw_ostream&, llvm::String
              Ref, bool)':
  >> 12031    WithColor.cpp:(.text.unlikely._ZN4llvm9WithColor7warningERNS_11raw_ostreamENS_9StringRefEb+0x4c): undefined reference to `_intel_fast_memcpy'
     12032    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/oneapi-2025.0.4/binutils-2.43.1-6o5ttv7sqyrmss7z7yhjxw22pubvmuu7/bin/ld: /tmp/root/spack-stage/spack-stage-llvm-19.1.7
              -ggylpyfhgdtne32b4cqxsqpmmdyngqji/spack-build-ggylpyf/lib/libLLVMSupport.a(WithColor.cpp.o): in function `llvm::WithColor::note(llvm::raw_ostream&, llvm::StringRef
              , bool)':
...
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 1.0.0.dev0 (ab296da971506fb9626a2baa9311195889929069)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-zen4

### Additional information

[spack-build-env.txt](https://github.com/user-attachments/files/19375354/spack-build-env.txt)
[spack-build-out.txt](https://github.com/user-attachments/files/19375355/spack-build-out.txt)

@haampie @skosukhin @trws @PiotrSacharuk @rscohn2 

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

A similar issue seems to be blocking my PR #47642 (see e.g. this CI failure https://gitlab.spack.io/spack/spack/-/jobs/15220744), so I looked a bit into the log files posted in this issue. Here's what I found:

Looking at the compilation output: there are at least two sources for these "undefined reference to" errors:
* External dependency on `libffi.a` which is built with the Intel compiler and needs one of the symbols `_intel_fast_*`.
* LLVM source files that are compiled only at the first stage using the Intel compiler (e.g. `MicrosoftDemangleNodes.cpp`) and then put in `*.a` archive (e.g. `MicrosoftDemangleNodes.cpp.o` goes into `libLLVMDemangle.a`). These archives are later (during stage 3 of the LLVM build I think) used on link lines that use the newly built `clang++` which does not know where to find the missing symbols `_intel_fast_*` generated by the Intel compiler when compiling sources like `MicrosoftDemangleNodes.cpp`.

One simple solution may be to explicitly add the Intel library that contains the missing symbols (`libirc.a` or `libirc.so` found in `<intel-prefix>/lib`) to `LDFLAGS` when configuring LLVM with CMake.

> [@eugeneswalker](https://github.com/eugeneswalker) [@haampie](https://github.com/haampie) [@skosukhin](https://github.com/skosukhin) [@trws](https://github.com/trws) [@PiotrSacharuk](https://github.com/PiotrSacharuk) [@rscohn2](https://github.com/rscohn2) JFYI, I have same error in my end. In my case it is with `oneapi-2024.2.1`. I wonder if you have any workaround? Maybe downgrading to older version of llvm would work but I did not try.

Downgrading doesn't help. We have configurations with lower version and we see the same


> A similar issue seems to be blocking my PR [#47642](https://github.com/spack/spack/pull/47642) (see e.g. this CI failure https://gitlab.spack.io/spack/spack/-/jobs/15220744), so I looked a bit into the log files posted in this issue. Here's what I found:
> 
> Looking at the compilation output: there are at least two sources for these "undefined reference to" errors:
> 
> * External dependency on `libffi.a` which is built with the Intel compiler and needs one of the symbols `_intel_fast_*`.
> * LLVM source files that are compiled only at the first stage using the Intel compiler (e.g. `MicrosoftDemangleNodes.cpp`) and then put in `*.a` archive (e.g. `MicrosoftDemangleNodes.cpp.o` goes into `libLLVMDemangle.a`). These archives are later (during stage 3 of the LLVM build I think) used on link lines that use the newly built `clang++` which does not know where to find the missing symbols `_intel_fast_*` generated by the Intel compiler when compiling sources like `MicrosoftDemangleNodes.cpp`.
> 
> One simple solution may be to explicitly add the Intel library that contains the missing symbols (`libirc.a` or `libirc.so` found in `<intel-prefix>/lib`) to `LDFLAGS` when configuring LLVM with CMake.

I think it is something around you mentioned. Yesterday I changed probably too much but near the LDFLAGS. llvm built but after that I saw few problems with linking other packages used llvm (mesa). On Monday I will try to limit changes only to part which you suggest. Thank you for sharing your observations.

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -175,9 +175,9 @@ spack:
   # --
   # - paraview +qt                  # concretize: paraview: Qt support requires GLX on non Windows; llvm-17.0.6: https://github.com/spack/spack/issues/49625
   - py-cinemasci                  # llvm-14.0.6: https://github.com/spack/spack/issues/49625
-  - visit                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
+  # - visit                         # vtk-9.2.6: ??
   # --
-  - chapel ~cuda ~rocm            # llvm-19.1.7: https://github.com/spack/spack/issues/49625
+  # - chapel ~cuda ~rocm            # chapel-2.4.0: KeyError: 'intel-oneapi-compilers': /builds/spack/spack/var/spack/repos/builtin/packages/chapel/package.py:602, in setup_chpl_compilers: env.set("CHPL_HOST_COMPILER", self.compiler_map[self.spec.compiler.name])
   # - cp2k +mpi                     # dbcsr-2.8.0: FAILED: src/CMakeFiles/dbcsr.dir/dbcsr_api.F-pp.f src/CMakeFiles/dbcsr.dir/dbcsr_api.F.o.ddi: 
   # - dealii                        # taskflow@3.7.0: cmake: Taskflow currently supports the following compilers: g++ v7.0 or above, clang++ v6.0 or above
   # - exago +mpi ~ipopt +hiop ~python +raja ^hiop+raja~sparse # raja-0.14.0: RAJA/pattern/kernel/Tile.hpp:174:30: error: no member named 'block_id' in 'IterableTiler<Iterable>'
@@ -186,7 +186,7 @@ spack:
   # - fpm                           # fpm-0.10.0: /tmp/ifx1305151083OkWTRB/ifxqBG60i.i90: error #6405: The same named entity from different modules and/or program units cannot be referenced.   [TOML_TABLE]; fpm.F90(32048): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FPM_MANIFEST_PREPROCESS]
   # - geopm-runtime                 # concretize: c-blosc2: conflicts with '%oneapi';
   - glvis                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
-  - gptune ~mpispawn              # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  # - gptune ~mpispawn              # py-maturin-1.8.3: rust-lld: error: undefined symbol: __intel_cpu_feature_indicator_x
   # - lbann                         # lbann-0.104: https://github.com/spack/spack/issues/49619
   # - libpressio +bitgrooming +bzip2 ~cuda ~cusz +fpzip +hdf5 +libdistributed +lua +openmp +python +sz +sz3 +unix +zfp # concretize: c-blosc2: conflicts with '%oneapi';
   # - nek5000 +mpi ~visit           # nek5000-19.0: RuntimeError: Cannot build example: short_tests/eddy.

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -174,7 +174,7 @@ spack:
   - zfp
   # --
   # - paraview +qt                  # concretize: paraview: Qt support requires GLX on non Windows; llvm-17.0.6: https://github.com/spack/spack/issues/49625
-  - py-cinemasci                  # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  # - py-cinemasci                  # py-maturin-1.8.3: rust-lld: error: undefined symbol: _intel_fast_memcpy
   # - visit                         # vtk-9.2.6: ??
   # --
   # - chapel ~cuda ~rocm            # chapel-2.4.0: KeyError: 'intel-oneapi-compilers': /builds/spack/spack/var/spack/repos/builtin/packages/chapel/package.py:602, in setup_chpl_compilers: env.set("CHPL_HOST_COMPILER", self.compiler_map[self.spec.compiler.name])

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -173,7 +173,7 @@ spack:
   - vtk-m ~openmp                   # +openmp: https://github.com/spack/spack/issues/31830
   - zfp
   # --
-  - paraview +qt                  # llvm-17.0.6: https://github.com/spack/spack/issues/49625
+  # - paraview +qt                  # concretize: paraview: Qt support requires GLX on non Windows; llvm-17.0.6: https://github.com/spack/spack/issues/49625
   - py-cinemasci                  # llvm-14.0.6: https://github.com/spack/spack/issues/49625
   - visit                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
   # --

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -173,11 +173,11 @@ spack:
   - vtk-m ~openmp                   # +openmp: https://github.com/spack/spack/issues/31830
   - zfp
   # --
-  # - paraview +qt                  # llvm-17.0.6: https://github.com/spack/spack/issues/49625
-  # - py-cinemasci                  # llvm-14.0.6: https://github.com/spack/spack/issues/49625
-  # - visit                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
+  - paraview +qt                  # llvm-17.0.6: https://github.com/spack/spack/issues/49625
+  - py-cinemasci                  # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  - visit                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
   # --
-  # - chapel ~cuda ~rocm            # llvm-19.1.7: https://github.com/spack/spack/issues/49625
+  - chapel ~cuda ~rocm            # llvm-19.1.7: https://github.com/spack/spack/issues/49625
   # - cp2k +mpi                     # dbcsr-2.8.0: FAILED: src/CMakeFiles/dbcsr.dir/dbcsr_api.F-pp.f src/CMakeFiles/dbcsr.dir/dbcsr_api.F.o.ddi: 
   # - dealii                        # taskflow@3.7.0: cmake: Taskflow currently supports the following compilers: g++ v7.0 or above, clang++ v6.0 or above
   # - exago +mpi ~ipopt +hiop ~python +raja ^hiop+raja~sparse # raja-0.14.0: RAJA/pattern/kernel/Tile.hpp:174:30: error: no member named 'block_id' in 'IterableTiler<Iterable>'
@@ -185,8 +185,8 @@ spack:
   # - fftx                          # fftx-1.2.0: https://github.com/spack/spack/issues/49621
   # - fpm                           # fpm-0.10.0: /tmp/ifx1305151083OkWTRB/ifxqBG60i.i90: error #6405: The same named entity from different modules and/or program units cannot be referenced.   [TOML_TABLE]; fpm.F90(32048): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FPM_MANIFEST_PREPROCESS]
   # - geopm-runtime                 # concretize: c-blosc2: conflicts with '%oneapi';
-  # - glvis                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
-  # - gptune ~mpispawn              # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  - glvis                         # llvm-17.0.6: https://github.com/spack/spack/issues/49625
+  - gptune ~mpispawn              # llvm-14.0.6: https://github.com/spack/spack/issues/49625
   # - lbann                         # lbann-0.104: https://github.com/spack/spack/issues/49619
   # - libpressio +bitgrooming +bzip2 ~cuda ~cusz +fpzip +hdf5 +libdistributed +lua +openmp +python +sz +sz3 +unix +zfp # concretize: c-blosc2: conflicts with '%oneapi';
   # - nek5000 +mpi ~visit           # nek5000-19.0: RuntimeError: Cannot build example: short_tests/eddy.
@@ -209,8 +209,8 @@ spack:
   # --
   # - py-jupyterlab                 # py-maturin: rust-lld: error: undefined symbol: _intel_fast_memcpy
   # - py-notebook                   # py-maturin: rust-lld: error: undefined symbol: _intel_fast_memcpy
-  # - py-numba                      # llvm-14.0.6: https://github.com/spack/spack/issues/49625
-  # - py-pandas                     # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  - py-numba                      # llvm-14.0.6: https://github.com/spack/spack/issues/49625
+  - py-pandas                     # llvm-14.0.6: https://github.com/spack/spack/issues/49625
   # - py-plotly                     # py-maturin: rust-lld: error: undefined symbol: _intel_fast_memcpy
 
   - aml +level_zero

--- a/var/spack/repos/builtin/packages/llvm/package.py
+++ b/var/spack/repos/builtin/packages/llvm/package.py
@@ -951,6 +951,13 @@ class Llvm(CMakePackage, CudaPackage, LlvmDetection, CompilerPackage):
             # set the SDKROOT so the bootstrap compiler finds its C++ headers
             env.set("SDKROOT", macos_sdk_path())
 
+        if self.spec.satisfies("%intel-oneapi-compilers"):
+            intel_libs = find_libraries(
+                ["libsvml", "libimf", "libirc"],
+                self.spec["intel-oneapi-runtime"].prefix.lib
+            )
+            env.append_flags("LDFLAGS", intel_libs.ld_flags)
+
     def setup_run_environment(self, env: EnvironmentModifications) -> None:
         if self.spec.satisfies("+clang"):
             env.set("CC", join_path(self.spec.prefix.bin, "clang"))

--- a/var/spack/repos/builtin/packages/llvm/package.py
+++ b/var/spack/repos/builtin/packages/llvm/package.py
@@ -953,8 +953,7 @@ class Llvm(CMakePackage, CudaPackage, LlvmDetection, CompilerPackage):
 
         if self.spec.satisfies("%intel-oneapi-compilers"):
             intel_libs = find_libraries(
-                ["libsvml", "libimf", "libirc"],
-                self.spec["intel-oneapi-runtime"].prefix.lib
+                ["libsvml", "libimf", "libirc"], self.spec["intel-oneapi-runtime"].prefix.lib
             )
             env.append_flags("LDFLAGS", intel_libs.ld_flags)
 

